In [60]:
# Project 3

import plotly.io as pio

pio.renderers.default = "vscode+jupyterlab+notebook_connected"

plotly.io allows to save a Plotly figure as an image or an HTML file. This will be important when I go to publish this work on this website. A plotly figure can be data visualization in the form of a chart or graph. 

In [61]:
import pandas as pd 
import plotly.express as px

pandas and plotly.express are packages with native python functions. You can think of them like tool boxes full of useful tools needed to perform a variety for jobs.
For example, later on in this webpage you will see that I calculated the mean of my data by writing ".mean()" at the end of the line of code.
".mean()" is a function that I did not have to create, but a function that already existed in the package called pandas. 

In [62]:
GBV = pd.read_csv("ENDGBV__The_Intersection_of_Domestic_Violence__Race_Ethnicity_and_Sex_20241207.csv")
GBV

C:\Users\Diane\AppData\Local\Temp\ipykernel_5380\1006744794.py:1: DtypeWarning:

Columns (5,6,8,9) have mixed types. Specify dtype option on import or set low_memory=False.



Offense Type Report Date  Incident Precinct Code Borough Name  \
0       FELONY ASSAULT  07/15/2020                      52        BRONX   
1       FELONY ASSAULT  04/17/2020                     113       QUEENS   
2       FELONY ASSAULT  03/19/2020                      41        BRONX   
3       FELONY ASSAULT  04/29/2020                      45        BRONX   
4       FELONY ASSAULT  10/18/2020                      47        BRONX   
...                ...         ...                     ...          ...   
483785             DIR  04/30/2021                     109       QUEENS   
483786             DIR  10/30/2020                      42        BRONX   
483787             DIR  11/07/2020                      42        BRONX   
483788             DIR  12/04/2020                      47        BRONX   
483789             DIR  06/15/2020                      94     BROOKLYN   

       Intimate Relationship Flag     Victim Race Victim Sex  \
0                             Yes  WHITE HISPANIC       MALE   
1                             Yes  WHITE HISPANIC     FEMALE   
2                              No  WHITE HISPANIC       MALE   
3                              No  WHITE HISPANIC       MALE   
4                             Yes  WHITE HISPANIC       MALE   
...                           ...             ...        ...   
483785                    MISSING             NaN        NaN   
483786                    MISSING             NaN        NaN   
483787                    MISSING             NaN        NaN   
483788                    MISSING             NaN        NaN   
483789                    MISSING             NaN        NaN   

        Victim Reported Age Suspect Race Suspect Sex  Suspect Reported Age  \
0                       NaN          NaN      FEMALE                   NaN   
1                       NaN          NaN        MALE                   NaN   
2                       3.0        BLACK        MALE                   NaN   
3                       5.0        WHITE        MALE                   NaN   
4                       5.0        BLACK        MALE                   NaN   
...                     ...          ...         ...                   ...   
483785                  NaN          NaN         NaN                   NaN   
483786                  NaN          NaN         NaN                   NaN   
483787                  NaN          NaN         NaN                   NaN   
483788                  NaN          NaN         NaN                   NaN   
483789                  NaN          NaN         NaN                   NaN   

        COMMDIST  Poverty  Median Income  Unemployment  
0          207.0      1.0            1.0           1.0  
1          412.0      NaN            NaN           1.0  
2          202.0      1.0            1.0           1.0  
3          210.0      NaN            NaN           NaN  
4          212.0      NaN            NaN           1.0  
...          ...      ...            ...           ...  
483785       NaN      NaN            NaN           NaN  
483786       NaN      NaN            NaN           NaN  
483787       NaN      NaN            NaN           NaN  
483788       NaN      NaN            NaN           NaN  
483789       NaN      NaN            NaN           NaN  

[483790 rows x 15 columns]

In the above, I loaded the data that I wanted to use into Jupyter Notebook. Jupyter Notebook is the platform I used to code this project. The code returns a "preview" of the data for me. There is a dtype warning indicating that there are different dtypes in this CSV. A dtype can be "float", "integer", or "object" to name a few. "Float" is a number with a decimal, "integer" is a whole number without a decimal, and "object" can be a string of letters or words. You will see some values called NaN which means "not a number". NaN is a placeholder for missing values or not applicable values as having zeroes in my data set can skew my data analysis. The dtype warning is nothing to be worried about just yet. In the next few lines of code, I will look through the values I want to analyze to see if I have to convert any values so that they are all the same dtypes before I start manipulating the data. Having different dtypes is not always a bad thing when it comes to data analysis, you have to use your better judgement in whether you need to convert or drop values. 

In [63]:
GBV.head(10)

Offense Type Report Date  Incident Precinct Code   Borough Name  \
0  FELONY ASSAULT  07/15/2020                      52          BRONX   
1  FELONY ASSAULT  04/17/2020                     113         QUEENS   
2  FELONY ASSAULT  03/19/2020                      41          BRONX   
3  FELONY ASSAULT  04/29/2020                      45          BRONX   
4  FELONY ASSAULT  10/18/2020                      47          BRONX   
5  FELONY ASSAULT  10/18/2020                      47          BRONX   
6  FELONY ASSAULT  10/27/2021                      75       BROOKLYN   
7  FELONY ASSAULT  12/27/2021                     107         QUEENS   
8  FELONY ASSAULT  10/21/2021                     120  STATEN ISLAND   
9  FELONY ASSAULT  11/24/2021                      83       BROOKLYN   

  Intimate Relationship Flag     Victim Race Victim Sex  Victim Reported Age  \
0                        Yes  WHITE HISPANIC       MALE                  NaN   
1                        Yes  WHITE HISPANIC     FEMALE                  NaN   
2                         No  WHITE HISPANIC       MALE                  3.0   
3                         No  WHITE HISPANIC       MALE                  5.0   
4                        Yes  WHITE HISPANIC       MALE                  5.0   
5                        Yes  WHITE HISPANIC       MALE                  5.0   
6                         No  WHITE HISPANIC       MALE                  8.0   
7                         No  WHITE HISPANIC       MALE                  8.0   
8                         No  WHITE HISPANIC     FEMALE                  8.0   
9                         No  WHITE HISPANIC     FEMALE                  9.0   

     Suspect Race Suspect Sex  Suspect Reported Age  COMMDIST  Poverty  \
0             NaN      FEMALE                   NaN     207.0      1.0   
1             NaN        MALE                   NaN     412.0      NaN   
2           BLACK        MALE                   NaN     202.0      1.0   
3           WHITE        MALE                   NaN     210.0      NaN   
4           BLACK        MALE                   NaN     212.0      NaN   
5  WHITE HISPANIC        MALE                   NaN     212.0      NaN   
6  WHITE HISPANIC        MALE                   NaN     305.0      1.0   
7  WHITE HISPANIC         NaN                   NaN     408.0      NaN   
8             NaN        MALE                   NaN     501.0      NaN   
9  WHITE HISPANIC        MALE                   NaN     304.0      1.0   

   Median Income  Unemployment  
0            1.0           1.0  
1            NaN           1.0  
2            1.0           1.0  
3            NaN           NaN  
4            NaN           1.0  
5            NaN           1.0  
6            1.0           NaN  
7            NaN           NaN  
8            NaN           NaN  
9            1.0           NaN

.head() returns the first five rows of a data set. I wrote in 10 to see the first 10 rows.

In [64]:
GBV.tail(10)

Offense Type Report Date  Incident Precinct Code Borough Name  \
483780          DIR  07/05/2021                      50        BRONX   
483781          DIR  07/05/2021                      50        BRONX   
483782          DIR  09/10/2021                      61     BROOKLYN   
483783          DIR  09/22/2021                      63     BROOKLYN   
483784          DIR  06/17/2021                      94     BROOKLYN   
483785          DIR  04/30/2021                     109       QUEENS   
483786          DIR  10/30/2020                      42        BRONX   
483787          DIR  11/07/2020                      42        BRONX   
483788          DIR  12/04/2020                      47        BRONX   
483789          DIR  06/15/2020                      94     BROOKLYN   

       Intimate Relationship Flag Victim Race Victim Sex  Victim Reported Age  \
483780                    MISSING         NaN        NaN                  NaN   
483781                    MISSING         NaN        NaN                  NaN   
483782                    MISSING         NaN        NaN                  NaN   
483783                    MISSING         NaN        NaN                  NaN   
483784                    MISSING         NaN        NaN                  NaN   
483785                    MISSING         NaN        NaN                  NaN   
483786                    MISSING         NaN        NaN                  NaN   
483787                    MISSING         NaN        NaN                  NaN   
483788                    MISSING         NaN        NaN                  NaN   
483789                    MISSING         NaN        NaN                  NaN   

       Suspect Race Suspect Sex  Suspect Reported Age  COMMDIST  Poverty  \
483780          NaN         NaN                   NaN       NaN      NaN   
483781          NaN         NaN                   NaN       NaN      NaN   
483782          NaN         NaN                   NaN       NaN      NaN   
483783          NaN         NaN                   NaN       NaN      NaN   
483784          NaN         NaN                   NaN       NaN      NaN   
483785          NaN         NaN                   NaN       NaN      NaN   
483786          NaN         NaN                   NaN       NaN      NaN   
483787          NaN         NaN                   NaN       NaN      NaN   
483788          NaN         NaN                   NaN       NaN      NaN   
483789          NaN         NaN                   NaN       NaN      NaN   

        Median Income  Unemployment  
483780            NaN           NaN  
483781            NaN           NaN  
483782            NaN           NaN  
483783            NaN           NaN  
483784            NaN           NaN  
483785            NaN           NaN  
483786            NaN           NaN  
483787            NaN           NaN  
483788            NaN           NaN  
483789            NaN           NaN

.tail() returns the last five rows of a data set. I wrote in 10 to see the last 10 rows. 

In [65]:
GBV.sample(10)

Offense Type Report Date  Incident Precinct Code   Borough Name  \
171952          DIR  05/29/2020                      32      MANHATTAN   
480485          DIR  01/01/2020                     106         QUEENS   
274515          DIR  08/16/2020                     101         QUEENS   
441904          DIR  04/18/2021                      68       BROOKLYN   
119757          DIR  05/25/2021                      47          BRONX   
83836           DIR  04/27/2020                      19      MANHATTAN   
134503          DIR  09/17/2020                     112         QUEENS   
327092          DIR  12/07/2021                     120  STATEN ISLAND   
231192          DIR  07/19/2021                      46          BRONX   
336938          DIR  01/22/2021                      52          BRONX   

       Intimate Relationship Flag Victim Race Victim Sex  Victim Reported Age  \
171952                         NO       BLACK     FEMALE                  NaN   
480485                    MISSING         NaN        NaN                  NaN   
274515                         NO       BLACK     FEMALE                 76.0   
441904                    MISSING         NaN        NaN                  NaN   
119757                        YES       WHITE     FEMALE                 51.0   
83836                         YES       WHITE     FEMALE                 34.0   
134503                    MISSING       OTHER     FEMALE                  NaN   
327092                         NO       BLACK     FEMALE                 71.0   
231192                        YES       BLACK     FEMALE                 20.0   
336938                         NO       BLACK       MALE                 46.0   

       Suspect Race Suspect Sex  Suspect Reported Age  COMMDIST  Poverty  \
171952        BLACK        MALE                   NaN     110.0      1.0   
480485          NaN         NaN                   NaN     410.0      NaN   
274515        BLACK        MALE                  29.0     414.0      NaN   
441904          NaN         NaN                   NaN     310.0      NaN   
119757        WHITE        MALE                  61.0     212.0      NaN   
83836         OTHER        MALE                  37.0     108.0      NaN   
134503          NaN         NaN                   NaN     406.0      NaN   
327092        BLACK        MALE                  42.0     501.0      NaN   
231192        WHITE        MALE                  20.0     205.0      1.0   
336938        BLACK        MALE                  47.0     207.0      1.0   

        Median Income  Unemployment  
171952            1.0           1.0  
480485            NaN           NaN  
274515            NaN           NaN  
441904            NaN           NaN  
119757            NaN           1.0  
83836             NaN           NaN  
134503            NaN           NaN  
327092            NaN           NaN  
231192            1.0           1.0  
336938            1.0           1.0

.sample() shows you a random assortment of rows from your data set. I also wrote in 10 here so that it would generate 10 random rows from my data set. 

Dataset to be used: https://data.cityofnewyork.us/Public-Safety/ENDGBV-The-Intersection-of-Domestic-Violence-Race-/2rb7-7eqa/about_data

Analysis question: Do incidents of gender-based violence (GBV) increase as the age of the victim decreases? I am asking if there is an inverse relationship between incidents of violence and victims' ages.

Columns that will (likely) be used:

"Victim Reported Age"

"Number of Incidents" I will make a column in a new data frame called Number of Incidents for each age

Hypothesis: I hypothesize that age and frequency of domestic violence are not correlated. The data will not show a relationship between age and GBV incidents. I expect to see points that are scattered randomly with no apparent trend. 

Site URL: [the *.readthedocs.io URL of your live site, from the Publish section]

In [66]:
GBV.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 483790 entries, 0 to 483789
Data columns (total 15 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Offense Type                483790 non-null  object 
 1   Report Date                 483790 non-null  object 
 2   Incident Precinct Code      483790 non-null  int64  
 3   Borough Name                483790 non-null  object 
 4   Intimate Relationship Flag  483790 non-null  object 
 5   Victim Race                 387733 non-null  object 
 6   Victim Sex                  394987 non-null  object 
 7   Victim Reported Age         344524 non-null  float64
 8   Suspect Race                367160 non-null  object 
 9   Suspect Sex                 374345 non-null  object 
 10  Suspect Reported Age        300802 non-null  float64
 11  COMMDIST                    483686 non-null  float64
 12  Poverty                     202325 non-null  float64
 13  Median Income 

.info() returns all the columns, tells you the dtypes, as well as how many rows have values in the data set 

In [67]:
GBV["Victim Reported Age"].unique()

array([ nan,   3.,   5.,   8.,   9.,  10.,  11.,  12.,  13.,  14.,  15.,
        16.,  17.,  18.,  19.,  20.,  21.,  22.,  23.,  24.,  25.,  26.,
        27.,  28.,  29.,  30.,  31.,  32.,  33.,  34.,  35.,  36.,  37.,
        38.,  39.,  40.,  41.,  42.,  43.,  44.,  45.,  46.,  47.,  48.,
        49.,  50.,  51.,  52.,  53.,  54.,  55.,  56.,  58.,  59.,  60.,
        62.,  67.,  71.,  73.,  82.,  65.,   4.,   6.,   7.,  76.,  63.,
        68.,  74.,  57.,  70.,  75.,  83.,   1.,   2.,  64.,  69.,  78.,
        66.,  77.,  72.,  61.,  79.,  84.,  85.,  86.,  81.,  91.,  80.,
        88.,  89.,  87.,  90.,  92.,  93.,  94.,  95.,  96.,  97., 114.,
        99.,  98., 100., 101., 110., 105., 104., 120.])

.unique() in the way that I used it here shows how many unique values are in the column entitled "Victim Reported Age" in the data set. This is how you can find weird values that may not match with what you want to see from the specific column.  

In [68]:
GBV_list = GBV["Victim Reported Age"].unique()
GBV_list.sort()
list(GBV_list)

[1.0,
 2.0,
 3.0,
 4.0,
 5.0,
 6.0,
 7.0,
 8.0,
 9.0,
 10.0,
 11.0,
 12.0,
 13.0,
 14.0,
 15.0,
 16.0,
 17.0,
 18.0,
 19.0,
 20.0,
 21.0,
 22.0,
 23.0,
 24.0,
 25.0,
 26.0,
 27.0,
 28.0,
 29.0,
 30.0,
 31.0,
 32.0,
 33.0,
 34.0,
 35.0,
 36.0,
 37.0,
 38.0,
 39.0,
 40.0,
 41.0,
 42.0,
 43.0,
 44.0,
 45.0,
 46.0,
 47.0,
 48.0,
 49.0,
 50.0,
 51.0,
 52.0,
 53.0,
 54.0,
 55.0,
 56.0,
 57.0,
 58.0,
 59.0,
 60.0,
 61.0,
 62.0,
 63.0,
 64.0,
 65.0,
 66.0,
 67.0,
 68.0,
 69.0,
 70.0,
 71.0,
 72.0,
 73.0,
 74.0,
 75.0,
 76.0,
 77.0,
 78.0,
 79.0,
 80.0,
 81.0,
 82.0,
 83.0,
 84.0,
 85.0,
 86.0,
 87.0,
 88.0,
 89.0,
 90.0,
 91.0,
 92.0,
 93.0,
 94.0,
 95.0,
 96.0,
 97.0,
 98.0,
 99.0,
 100.0,
 101.0,
 104.0,
 105.0,
 110.0,
 114.0,
 120.0,
 nan]

I converted the array into a list because sometimes that makes the data easier to read. 

In [69]:
with pd.option_context("display.max_rows", 1000):
    display(GBV.groupby("Victim Reported Age").size())

Victim Reported Age
1.0        648
2.0        682
3.0        829
4.0       1033
5.0       1139
6.0       1140
7.0       1125
8.0       1037
9.0       1114
10.0      1157
11.0      1016
12.0      1267
13.0      1311
14.0      1565
15.0      1881
16.0      2302
17.0      2813
18.0      3792
19.0      4741
20.0      5785
21.0      6632
22.0      7662
23.0      8205
24.0      8545
25.0      9211
26.0      9717
27.0      9919
28.0     10599
29.0     10443
30.0     10818
31.0     10758
32.0     10256
33.0      9993
34.0      9806
35.0      9533
36.0      8897
37.0      8314
38.0      8478
39.0      8334
40.0      7912
41.0      7161
42.0      6935
43.0      6475
44.0      6044
45.0      5620
46.0      5109
47.0      4823
48.0      4803
49.0      4705
50.0      4770
51.0      4658
52.0      4230
53.0      4306
54.0      4230
55.0      4129
56.0      3860
57.0      3820
58.0      3552
59.0      3269
60.0      3100
61.0      2778
62.0      2458
63.0      2354
64.0      2095
65.0      1900
66.0 

.groupby() pulls certain that you want to see or work with from the data set. I want to work with age so I grouped the column "Victim Reported Age" and used .size() to give me the amount of times that age occurs in the data set. It is important to fully investigate your data so that you can determine how you want to clean your data before you go any further. From the above data investigation, I found that I did not need to convert dtypes. 

In [70]:
counts = GBV.groupby("Victim Reported Age").size().reset_index(name="Number of Incidents")
counts

Victim Reported Age  Number of Incidents
0                    1.0                  648
1                    2.0                  682
2                    3.0                  829
3                    4.0                 1033
4                    5.0                 1139
..                   ...                  ...
101                104.0                    1
102                105.0                    1
103                110.0                    1
104                114.0                    1
105                120.0                    1

[106 rows x 2 columns]

Same code as before but here I converted it into a dataframe. Pandas will add a index column on the very left that numbers the rows in the frame. I renamed the column that shows how many times an age occurs "Number of Incidents". 

In [71]:
fig = px.scatter(
    counts,
    x='Victim Reported Age',
    y='Number of Incidents',
    title='Scatter Plot of Victim Age vs. Number of Incidents 2020-2021',
    labels={'Victim Reported Age': 'Victim Age', 'Number of Incidents': 'Number of Incidents'},
    template='plotly'
)

fig.show()

This is where the plotly.express package comes in handy! The shorthand px stands for the plotly.express package. I gave plotly.express the data frame called counts to work from. Match up the columns of data for the x-axis and y-axis. As with any graph, you have to title it and label your axes. fig.show() then produces your graph to see!

In [72]:
spearman_corr = counts.corr(method='spearman')
print(f"Spearman Correlation: {spearman_corr}")

Spearman Correlation:                      Victim Reported Age  Number of Incidents
Victim Reported Age             1.000000            -0.656622
Number of Incidents            -0.656622             1.000000


I explain why I calculated the Spearman correlation further down in the webpage during my final analysis of the data. 

In [73]:
import numpy as np

# Extract the two variables
x = counts['Victim Reported Age']
y = counts['Number of Incidents']

# Calculate the correlation coefficient (Pearson by default)
correlation = np.corrcoef(x, y)[0, 1]

# Print the correlation value
print(f"Correlation Coefficient (Pearson): {correlation:.2f}")

Correlation Coefficient (Pearson): -0.51


I calculated the Pearson coefficient out of curiosity as well as to explain why I did not chose to use this coefficient. I also wanted to see how close it is to the Spearman coefficient. The Pearson coeffeicient is not useful for this data analysis because the scatterplot shows a non-linear relationship of the variables and the Pearson correlation coefficient better describes linear relationships. 

In [74]:
mode_value = GBV['Victim Reported Age'].mode()
print(f"Mode(s) of Victim Reported Age: {mode_value.tolist()}")

Mode(s) of Victim Reported Age: [30.0]


Like I previously explained, .mode(), .mean(), .min(), .max(), .median() are all native functions or "tools" of the pandas package. I am storing the information I get from the line of code in a variable with new name like "mode_value" and then having the mode returned to me by using the print function. You will see this same action in the next few lines of code.

In [75]:
max_value = GBV['Victim Reported Age'].max()
print(f"The maximum value in 'Victim Reported Age' is: {max_value}")

The maximum value in 'Victim Reported Age' is: 120.0


In [76]:
min_value = GBV['Victim Reported Age'].min()
print(f"The minimum value in 'Victim Reported Age' is: {min_value}")

The minimum value in 'Victim Reported Age' is: 1.0


In [77]:
mean_value = GBV['Victim Reported Age'].mean()
print(f"The mean value in 'Victim Reported Age' is: {mean_value}")

The mean value in 'Victim Reported Age' is: 37.193310190291534


In [78]:
median_value = GBV['Victim Reported Age'].median()
print(f"The median value in 'Victim Reported Age' is: {median_value}")

The median value in 'Victim Reported Age' is: 35.0


The scatterplot demonstrates enough evidence to reject my initial hypothesis. The data provided by NYC OpenData for years 2020 and 2021 shows a right-skewed distribution and quadratic relationship. The mean age for incidents of domestic violence will be higher than the median age for incidents of domestic violence. The scatter plot shows a weak negative correlation between the two variables with a correlation coefficient of -0.656622. I am using the Spearman correlation coefficient because it measures the monotonic relationship between two variables in non-linear data. What surprised me and disappointed me the most is seeing that incidents of GBV go as high as 120 years of age. Most of the data is centered around 20-40 years of age indicating that highest frequencies of domestic violence occur between 20-40 years of age. I did not distinguish between males and females so please bear in mind that this data analysis includes both genders. Columns in the data above show that domestic violence does happen to both genders. The scatterplot's peak is shown at age 30 which tells me that it is the mode for our data. Toggling over the chart, one can see that the number of incidents at the age of 30 totaled 10,818,000. I coded for the mode to cross-check my scatter plot which also gave 30 years of age. In future analysis, one should study the relationship between the incidents of domestic violence and victim's gender in NYC in the years of 2020 and 2021 to understand if the same relationship trends seen in this study still hold true. 

Research Question: Do incidents of gender-based violence (GBV) increase as the age of the victim decreases? I am asking if there is an inverse relationship between incidents of violence and victims' ages. 

Answer: Yes, the second half of the scatterplot shows that after the age of 30 the incidents of GBV have an inverse relationship with victims' age. The correlation coefficient of -0.656622 supports this conclusion. It is also important to note that because of the quadratic nature of the relationship between the variables that one can say that incidents of GBV increasing as age increases is also true.  